### PR TITLE
option to add extra volume tags with ebs csi driver

### PR DESCRIPTION
**Pull request by heybronson**
Adds `cloudConfig.awsEBSCSIDriver.extraVolumeTags` to add extra AWS tags to volumes provisioned by the EBS CSI Driver.

Changes the `join` template function to use the Sprig-defined function instead of `strings.Join` to get better type handling with the Sprig function. Updates other templates to fit the Sprig `join` expectations.

Testing:
`kops update cluster`:
```
Will modify resources:
  LaunchTemplate/master-us-west-2a.masters.bronson.cluster.doordash.cloud
        UserData
                                ...
                                      extraVolumeTags:
                                        testKey: testValue
                                +       testKey2: testValue2
                                      managed: true
                                      podAnnotations:
                                ...


  LaunchTemplate/nodes-us-west-2a.bronson.cluster.doordash.cloud
        UserData
                                ...
                                      extraVolumeTags:
                                        testKey: testValue
                                +       testKey2: testValue2
                                      managed: true
                                      podAnnotations:
                                ...


  LaunchTemplate/nodes-us-west-2b.bronson.cluster.doordash.cloud
        UserData
                                ...
                                      extraVolumeTags:
                                        testKey: testValue
                                +       testKey2: testValue2
                                      managed: true
                                      podAnnotations:
                                ...


  LaunchTemplate/nodes-us-west-2c.bronson.cluster.doordash.cloud
        UserData
                                ...
                                      extraVolumeTags:
                                        testKey: testValue
                                +       testKey2: testValue2
                                      managed: true
                                      podAnnotations:
                                ...


  ManagedFile/bronson.cluster.doordash.cloud-addons-aws-ebs-csi-driver.addons.k8s.io-k8s-1.17
        Contents
                                ...
                                          - --http-endpoint=0.0.0.0:3301
                                          - --v=5
                                +         - --extra-tags=testKey=testValue,testKey2=testValue2
                                -         - --extra-tags=testKey=testValue
                                          env:
                                          - name: CSI_ENDPOINT
                                ...


  ManagedFile/bronson.cluster.doordash.cloud-addons-bootstrap
        Contents
                                ...
                                    - id: k8s-1.17
                                      manifest: aws-ebs-csi-driver.addons.k8s.io/k8s-1.17.yaml
                                +     manifestHash: f84f76f9db04b6d51d42922182b570343fccf1d83f7a0cf1a257b68626f6776e
                                -     manifestHash: f2af46d8a54c397193b9d970fbe2777d8cdf94913d04eea943880d7216d7b0ee
                                      name: aws-ebs-csi-driver.addons.k8s.io
                                      selector:
                                ...


  ManagedFile/cluster-completed.spec
        Contents
                                ...
                                  metadata:
                                    creationTimestamp: "2022-05-11T22:35:23Z"
                                +   generation: 15
                                -   generation: 14
                                    name: bronson.cluster.doordash.cloud
                                  spec:
                                ...
                                        extraVolumeTags:
                                          testKey: testValue
                                +         testKey2: testValue2
                                        managed: true
                                        podAnnotations:
                                ...
```

`kubectl describe pod`:
```
❯ kubectl get deploy -n kube-system ebs-csi-controller -ojson | jq '.spec.template.spec.containers[0].args'
[
  "controller",
  "--endpoint=$(CSI_ENDPOINT)",
  "--logtostderr",
  "--k8s-tag-cluster-id=bronson.cluster.doordash.cloud",
  "--extra-tags=KubernetesCluster=bronson.cluster.doordash.cloud",
  "--http-endpoint=0.0.0.0:3301",
  "--v=5",
  "--extra-tags=testKey=testValue,testKey2=testValue2"
]
```

Confirmed tags are on new PVs:
image

### CODE REVIEW

File: upup/models/cloudup/resources/addons/aws-ebs-csi-driver.addons.k8s.io/k8s-1.17.yaml.template [hunk 1]
```diff
@@ -478,6 +478,13 @@ spec:
             - "--extra-tags={{ CloudLabels }}"
             - --http-endpoint=0.0.0.0:3301
             - --v=5
+            {{- $tags := list }}
+            {{- range $key, $value := .ExtraVolumeTags }}
+            {{- $tags = append $tags (printf "%s=%v" $key $value) }}
+            {{- end }}
+            {{- if gt (len $tags) 0 }}
+            {{ printf "- \"--extra-tags=%s\"" (join "," $tags) }}
```
Comment (hakman) [inline upup/models/cloudup/resources/addons/aws-ebs-csi-driver.addons.k8s.io/k8s-1.17.yaml.template:486]: @heybronson Any thought about just using `CloudLabels`?
See  line 478 above:
```
- "--extra-tags={{ CloudLabels }}"
```

File: upup/pkg/fi/cloudup/template_functions.go [hunk 2]
```diff
@@ -103,6 +100,11 @@ func (tf *TemplateFunctions) AddTo(dest template.FuncMap, secretStore fi.SecretS
 	dest["trimPrefix"] = sprigTxtFuncMap["trimPrefix"]
 	dest["semverCompare"] = sprigTxtFuncMap["semverCompare"]
 	dest["ternary"] = sprigTxtFuncMap["ternary"]
+	dest["dict"] = sprigTxtFuncMap["dict"]
+	dest["list"] = sprigTxtFuncMap["list"]
+	dest["set"] = sprigTxtFuncMap["set"]
+	dest["append"] = sprigTxtFuncMap["append"]
+	dest["join"] = sprigTxtFuncMap["join"]
```
Comment (olemarkus) [inline upup/pkg/fi/cloudup/template_functions.go:107]: It's probably best to do this as a separate PR or at least a separate commit.

File: upup/models/cloudup/resources/addons/aws-ebs-csi-driver.addons.k8s.io/k8s-1.17.yaml.template (hunk 1 above)
Comment (olemarkus) [reply upup/models/cloudup/resources/addons/aws-ebs-csi-driver.addons.k8s.io/k8s-1.17.yaml.template:486]: Because of the IAM policies it's also quite important that `CloudLabels` are unconditionally added.

**Review by olemarkus — CHANGES_REQUESTED** [review]

File: upup/models/cloudup/resources/addons/aws-ebs-csi-driver.addons.k8s.io/k8s-1.17.yaml.template (hunk 1 above)
Comment (heybronson) [reply upup/models/cloudup/resources/addons/aws-ebs-csi-driver.addons.k8s.io/k8s-1.17.yaml.template:486]: I like the idea of just using `CloudLabels`; that's the primary use case I was targeting anyways. Makes more sense to use the existing `CloudLabels` instead of defining them twice.

File: upup/pkg/fi/cloudup/template_functions.go (hunk 2 above)
Comment (heybronson) [reply upup/pkg/fi/cloudup/template_functions.go:107]: I'll make a separate PR to use sprig `join`.
